xmpp-vala/src/module/sasl.vala

```diff
@@ -44,110 +39,27 @@ namespace Xmpp.Sasl {
-            checksum.get_digest(res, ref SHA1_SIZE);
-            return res;
-        }
-
-        private static uint8[] hmac_sha1(uint8[] key, uint8[] data) {
-            Hmac hmac = new Hmac(ChecksumType.SHA1, key);
-            hmac.update(data);
-            uint8[] res = new uint8[SHA1_SIZE];
-            hmac.get_digest(res, ref SHA1_SIZE);
-            return res;
-        }
-
-        private static uint8[] pbkdf2_sha1(string password, uint8[] salt, uint iterations) {
-            uint8[] res = new uint8[SHA1_SIZE];
-            uint8[] last = new uint8[salt.length + 4];
-            for(int i = 0; i < salt.length; i++) {
-                last[i] = salt[i];
-            }
-            last[salt.length + 3] = 1;
-            for(int i = 0; i < iterations; i++) {
-                last = hmac_sha1((uint8[]) password.to_utf8(), last);
-                xor_inplace(res, last);
-            }
-            return res;
-        }
-
-        private static void xor_inplace(uint8[] mix, uint8[] a2) {
-            for(int i = 0; i < mix.length; i++) {
-                mix[i] = mix[i] ^ a2[i];
-            }
-        }
-
-        private static uint8[] xor(uint8[] a1, uint8[] a2) {
-            uint8[] mix = new uint8[a1.length];
-            for(int i = 0; i < a1.length; i++) {
-                mix[i] = a1[i] ^ a2[i];
-            }
-            return mix;
-        }
-
         public void received_nonza(XmppStream stream, StanzaNode node) {
-            if (node.ns_uri == NS_URI) {
-                if (node.name == "success") {
-                    Flag flag = stream.get_flag(Flag.IDENTITY);
-                    if (flag.mechanism == Mechanism.SCRAM_SHA_1) {
-                        string confirm = (string) Base64.decode(node.get_string_content());
-                        uint8[] server_signature = null;
-                        foreach(string c in confirm.split(",")) {
-                            string[] split = c.split("=", 2);
-                            if (split.length != 2) continue;
-                            switch(split[0]) {
-                                case "v": server_signature = Base64.decode(split[1]); break;
-                            }
-                        }
-                        if (server_signature == null) return;
-                        if (server_signature.length != flag.server_signature.length) return;
-                        for(int i = 0; i < server_signature.length; i++) {
-                            if (server_signature[i] != flag.server_signature[i]) return;
-                        }
-                    }
-                    stream.require_setup();
-                    flag.password = null; // Remove password from memory
-                    flag.finished = true;
-                } else if (node.name == "failure") {
+            if (node.ns_uri != NS_URI)
+                return;
+            if (node.name == "success") {
+                Flag flag = stream.get_flag(Flag.IDENTITY);
```

Comment:
I think with the new logic, the server signature (that is sent with the success nonza) is not verified any more (as described in points 10 and 11 in https://wiki.xmpp.org/web/SASL_and_SCRAM-SHA-1). Can you check how to verify it through gsasl?

Reply:
Done.